### ISSUE TITLE
Indexing operations in Einsum?

### ISSUE THREAD

**Issue by srush** [dgasmith/opt_einsum#75]
Hi opt-einsum, 

Big fan. I have a feature request / question.

We often find ourselves writing complicated indexing operations where we want something like `Z[i, j] = X[i, Y[i, j]]`. To do this in pytorch/numpy is a bug-prone hassle. 

Given that this is just a sparse matrix multiply, I love the idea of writing it in einsum, maybe something like `einsum(ik,ijk*->ij, x, y)`. Where k is the implied dimension of treating Y as a sparse one-hot matrix. 

Any ideas on how I might implement a backend like this or notate it? Would want the sparse dimension to become fancy indexing.

**Comment by dgasmith** [comment]
Hi, thanks for the interest!

This is an interesting operation. Can you give a small snippet of code where you accomplish this for the above example? Overall, I think we can do it. I am just not quite sure we can do it without python loops.

**Comment by srush** [comment]
Yeah, good point. 

In pytorch is becomes something like a view and an index_select. I guess that might require a different API though.

https://pytorch.org/docs/stable/torch.html#torch.index_select

**Comment by dgasmith** [comment]
Is this operation similar to [`np.take`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.take.html)? If so, we could likely create an `einsum` like interface to it.

**Comment by srush** [comment]
Yeah, this is identical to `np.take`, although I need to think about exactly how the semantics work here.

**Comment by srush** [comment]
So my example above was wrong and requires extra batching, but I think your intuition is correct. Simply replace `tensordot` with  `take/index_select` for index dimensions, and do them first,

`einsum("ki,ljk*->ijl", x, y.long())`

becomes 

`x.take(y, axis=2)` => "ki,ljk->lji"  

Also if someone really wants to sum out in an index tensor dim it maybe just become dense... 

`einsum("nk*->k", x, y.long())`

`eye(k, k).take(y, axis=1)` => "nk*->nk"

**Comment by dgasmith** [comment]
Can you provide a few working NumPy code examples?

What happens in this formalism when there are more than two tensors?
